File: src/Carbon Wrappers/XJKeyChain.m
#import "XJKeyChain.h"
#import <Security/Security.h>

static XJKeyChain* defaultKeyChain = nil;

@interface XJKeyChain ()

-(SecKeychainItemRef)_genericPasswordReferenceForService:(NSString *)service account:(NSString*)account CF_RETURNS_RETAINED;

@end

@implementation XJKeyChain
@synthesize maxPasswordLength;

+ (instancetype) defaultKeyChain {
	return ( defaultKeyChain ?: [[self alloc] init] );
}

- (instancetype)init
{
    self = [super init];
    maxPasswordLength = 127;
    return self;
}

- (OSStatus)setGenericPassword:(NSString*)password forService:(NSString *)service account:(NSString*)account
{
    OSStatus ret = noErr;
    SecKeychainItemRef itemref = NULL;
    void *p = (void *)alloca(maxPasswordLength * sizeof(char));
    
    if ([service length] == 0 || [account length] == 0) {
        return ret;
    }
    
    if (!password || [password length] == 0) {
        [self removeGenericPasswordForService:service account:account];
    } else {
        strcpy(p,[password UTF8String]);
    
        if ((itemref = [self _genericPasswordReferenceForService:service account:account]))
			SecKeychainItemDelete(itemref);
        SecKeychainRef keychain;
        SecKeychainCopyDefault(&keychain);
        ret = SecKeychainAddGenericPassword(keychain, (UInt32)[service lengthOfBytesUsingEncoding:NSUTF8StringEncoding], [service UTF8String], (UInt32)[account lengthOfBytesUsingEncoding:NSUTF8StringEncoding], [account UTF8String], (UInt32)strlen(p), p, &itemref);
        CFRelease(itemref);
    }
    return ret;
}

- (NSString*)genericPasswordForService:(NSString *)service account:(NSString*)account
{
    OSStatus ret = noErr;
    UInt32 length = 0;
    void *p = NULL;
    NSString *string = @"";
    
    if ([service length] == 0 || [account length] == 0) {
        return @"";
    }
    
    ret = SecKeychainFindGenericPassword(NULL, (UInt32)[service lengthOfBytesUsingEncoding:NSUTF8StringEncoding], [service UTF8String], (UInt32)[account lengthOfBytesUsingEncoding:NSUTF8StringEncoding], [account UTF8String], &length, &p, NULL);

    if (ret == noErr)
        string = [[NSString alloc] initWithBytes:p length:length encoding:NSUTF8StringEncoding];
	if (p) {
        SecKeychainItemFreeContent(NULL, p);
	}
    return string;
}

- (void)removeGenericPasswordForService:(NSString *)service account:(NSString*)account
{
    SecKeychainItemRef itemref = nil ;
    if ((itemref = [self _genericPasswordReferenceForService:service account:account]))
        SecKeychainItemDelete(itemref);
}

- (void)setMaxPasswordLength:(NSUInteger)length
{
    if (![self isEqual:defaultKeyChain]) {
        maxPasswordLength = length ;
    } else {
    }
}

- (SecKeychainItemRef)_genericPasswordReferenceForService:(NSString *)service account:(NSString*)account
{
    SecKeychainItemRef itemref = nil;
    SecKeychainFindGenericPassword(NULL, (UInt32)[service lengthOfBytesUsingEncoding:NSUTF8StringEncoding], [service UTF8String], (UInt32)[account lengthOfBytesUsingEncoding:NSUTF8StringEncoding], [account UTF8String], NULL, NULL, &itemref);
    return itemref;
}

@end

File: src/Carbon Wrappers/XJKeyChain.swift
//
//  XJKeyChain.swift
//  Xjournal
//
//  Created by C.W. Betts on 1/24/15.
//
//

import Foundation
import Security

private var adefaultKeyChain: KeyChain = {
	return KeyChain()
}()

final class KeyChain: NSObject {
	var maxPasswordLength: UInt32
	private let currentKeychain: SecKeychain?
	
	enum PreferencesDomain: UInt32 {
		///user domain
		case User
		
		///system (daemon) domain
		case System
		
		///preferences to be merged to everyone
		case Common
		
		///dynamic searchlist (typically removable keychains like smartcards)
		case Dynamic
	}

	required convenience override init() {
		var defKeyChain: SecKeychain?
		_ = SecKeychainCopyDefault(&defKeyChain)
		self.init(keychain: defKeyChain)
	}
	
	init(keychain: SecKeychain?) {
		maxPasswordLength = 127;
		currentKeychain = keychain
		
		super.init()
	}
	
	convenience init?(keychainPath: NSURL) {
		var capturedKey: SecKeychain?
		
		_ = SecKeychainOpen(keychainPath.fileSystemRepresentation, &capturedKey)
		if let aKey = capturedKey {
			self.init(keychain: aKey)
		} else {
			self.init(keychain: nil)
			
			return nil
		}
	}
	
	@objc(defaultKeyChain) static let `default`: KeyChain = KeyChain()
	
	private func genericPasswordReference(service: String, account: String) -> SecKeychainItem? {
		var itemref: SecKeychainItem? = nil
		SecKeychainFindGenericPassword(currentKeychain, UInt32(service.lengthOfBytes(using: String.Encoding.utf8)), service, UInt32(account.lengthOfBytes(using: String.Encoding.utf8)), account, nil, nil, &itemref)
		
		return itemref
	}
	
	@objc(removeGenericPasswordForService:account:) func removeGenericPassword(service: String, account: String) {
		if let itemref = genericPasswordReference(service: service, account: account) {
			SecKeychainItemDelete(itemref)
		}
	}
	
	@objc(setGenericPassword:forService:account:) func setGenericPassword(password: String?, service: String, account: String) {
		//var ret: OSStatus = noErr
		
		if service.characters.count == 0 || account.characters.count == 0 {
			return
		}
		
		if let aPass = password {
			if let itemRef = genericPasswordReference(service: service, account: account) {
				SecKeychainItemDelete(itemRef)
			}
			SecKeychainAddGenericPassword(currentKeychain, UInt32(service.lengthOfBytes(using: String.Encoding.utf8)), service, UInt32(account.lengthOfBytes(using: String.Encoding.utf8)), account, UInt32(aPass.lengthOfBytes(using: String.Encoding.utf8)), UnsafeRawPointer((aPass as NSString).utf8String!), nil)
		} else {
			removeGenericPassword(service: service, account: account)
		}
	}
	
	@objc(genericPasswordForService:account:) func genericPassword(service: String, account: String) -> String {
		var ret: OSStatus = noErr
		var string = ""
		var p: UnsafeMutableRawPointer? = nil
		var length: UInt32 = 0
		
		if service.characters.count == 0 || account.characters.count == 0 {
			return ""
		}

		ret = SecKeychainFindGenericPassword(currentKeychain, UInt32(service.lengthOfBytes(using: String.Encoding.utf8)), service, UInt32(account.lengthOfBytes(using: String.Encoding.utf8)), account, &length, &p, nil)
		
		if ret == noErr {
			string = NSString(bytes: p!, length: Int(length), encoding: String.Encoding.utf8.rawValue) as! String
		}
		if p != nil {
			SecKeychainItemFreeContent(nil, p)
		}
		
		return string
	}
	
	var keychainURL: URL? {
		var pathLen: UInt32 = 0
		var pathName: [Int8] = [Int8](repeating: 0, count: Int(PATH_MAX))
		let iErr = SecKeychainGetPath(currentKeychain, &pathLen, &pathName)
		if iErr != noErr {
			return nil
		}
		pathName[Int(pathLen)] = 0
		
		return URL(fileURLWithFileSystemRepresentation: pathName, isDirectory: false, relativeTo: nil)
	}
}
